### ISSUE TITLE
Terraform Init S3 Permission denied, not getting proper role from GetCallerIdentity

### ISSUE THREAD

**Issue by mjgorman** [hashicorp/terraform#26237]
### Terraform Version

```
Terraform v0.12.29
```

### Terraform Configuration Files

```terraform
provider "aws" {
  region  = var.region
}

terraform {
  backend "s3" {
    bucket = "my-companies-bucket"
    key    = "prod-east-1/terraform.tfstate"
    region = "us-east-1"
    encrypt = true
  }
}

```

### Debug Output

Meaningful bit of the init output showing the incorrect role being returned by GetCallerIdentity. 
```
-----------------------------------------------------
2020/09/14 16:15:50 [DEBUG] [aws-sdk-go] DEBUG: Response sts/GetCallerIdentity Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK
Connection: close
Content-Length: 470
Content-Type: text/xml
Date: Mon, 14 Sep 2020 16:15:49 GMT
X-Amzn-Requestid: aa451b13-5a3d-448f-8342-ade3cf7ab925


-----------------------------------------------------
2020/09/14 16:15:50 [DEBUG] [aws-sdk-go] <GetCallerIdentityResponse xmlns="https://sts.amazonaws.com/doc/2011-06-15/">
  <GetCallerIdentityResult>
    <Arn>arn:aws:sts::ACCOUNT_NUMBER:assumed-role/dev-primary-workers/i-022fe44e1e0e1a9eb</Arn>
    <UserId>REDACTED</UserId>
    <Account>ACCOUNT_NUMBER</Account>
  </GetCallerIdentityResult>
  <ResponseMetadata>
    <RequestId>aa451b13-5a3d-448f-8342-ade3cf7ab925</RequestId>
  </ResponseMetadata>
</GetCallerIdentityResponse>
```

AWS Cli output from get-Caller-identity on the same kubernetes pod.
```
/home/jenkins/agent/workspace/terraform/terraform/prod-east-1 # aws sts get-caller-identity
{
    "UserId": "REDACTED",
    "Account": "ACCOUNT_NUMBER",
    "Arn": "arn:aws:sts::ACCOUNT_NUMBER:assumed-role/cicd-k8s-service-role/botocore-session-1600100008"
}
```

### Expected Behavior

When using Terraform from within a Kubernetes pod which uses ServiceRole to map IAM roles to pods Terraform should discover the proper role to access amazon resources such as the s3 backend.

### Actual Behavior

The default role from the server is being used instead of the pod set service role, the caller identity should be the same in terraform as it is in the aws cli run on the same pod. 

### Steps to Reproduce

Create a pod with the following configuration, a matching configuration for serviceAccount should exist in the cluster and namespace matching the one speicfied and be mapped to an AWS role. 
```
apiVersion: v1
kind: Pod
spec:
  serviceAccountName: cicd
  hostIPC: true
  securityContext:
    fsGroup: 1000
  containers:
  - name: terraform
    image: hashicorp/terraform:0.12.29
    resources:
      requests:
        cpu: "500m"
        memory: "512Mi"
      limits:
        cpu: "500m"
        memory: "512Mi"
    command:
    - cat
    tty: true
```
Then run an init with an s3 backend. 
```
terraform init
```
You'll see a permission denied. 
```
Initializing the backend...


Successfully configured the backend "s3"! Terraform will automatically

use this backend unless the backend configuration changes.

Error refreshing state: AccessDenied: Access Denied

	status code: 403, request id: aaa, host id: bbb

```
Verified that aws CLI is detecting the proper role be default in the container from pod/instance metadata. 
```
aws sts get-caller-identity
```

### Additional Context

Running in a Kubernetes pod from Jenkins

**Comment by mjgorman** [comment]
Did some further debugging. 

The AWS env vars for role are set:
```
/home/jenkins/agent/workspace/terraform/terraform/dev-east-1 # env | grep AWS
AWS_ROLE_ARN=arn:aws:iam::ACCOUNT_NUMBER:role/cicd-k8s-service-role
AWS_WEB_IDENTITY_TOKEN_FILE=/var/run/secrets/eks.amazonaws.com/serviceaccount/token
```

The token file is also readable by the running user, so the sdk shouldn't have any issues loading those to do a proper assume unless there is some logic missing somewhere compared to the other versions of the sdk and the cli itself.

**Comment by bflad** [comment]
Hi @mjgorman 👋  Thank you for reporting this and sorry you ran into trouble here.

This appears to be similar to https://github.com/hashicorp/terraform/issues/24165 and is likely fixed by Terraform 0.13 and later due to https://github.com/hashicorp/terraform/pull/25134 which adjusts the S3 Backend authentication ordering to be more similar to the default AWS Go SDK ordering, ensuring EKS credentials are read before any EC2 Metadata.

There are some potential workarounds on Terraform 0.12.29, such as adding the `skip_metadata_api_check = true` configuration, if upgrading is not immediately an option.
